app/controllers/discourse_subscriptions/subscribe_controller.rb

```diff
@@ -53,24 +53,31 @@ def create
       begin
         customer = create_customer(params[:source])
         plan = ::Stripe::Price.retrieve(params[:plan])
+        promo_code = ::Stripe::PromotionCode.list({ code: params[:promo] }) if params[:promo].present?
+        promo_code = promo_code[:data][0] if promo_code && promo_code[:data] # we assume promo codes have a unique name
 
         recurring_plan = plan[:type] == 'recurring'
 
         if recurring_plan
           trial_days = plan[:metadata][:trial_period_days] if plan[:metadata] && plan[:metadata][:trial_period_days]
 
+          promo_code_id = promo_code[:id] if promo_code
+
           transaction = ::Stripe::Subscription.create(
             customer: customer[:id],
             items: [{ price: params[:plan] }],
             metadata: metadata_user,
-            trial_period_days: trial_days
+            trial_period_days: trial_days,
+            promotion_code: promo_code_id
           )
 
           payment_intent = retrieve_payment_intent(transaction[:latest_invoice]) if transaction[:status] == 'incomplete'
         else
+          coupon_id = promo_code[:coupon][:id] if promo_code && promo_code[:coupon] && promo_code[:coupon][:id]
```

Comment:
I think a bit cleaner way of doing this is 
```
coupon_id = promo_code.fetch(:coupon, {}).fetch(:id, {})
```

Reply:
Unfortunately, Stripe doesn't return a Hash object. Could look into converting a hash but not sure how that affects data processing in other ways...